src/axl.c

```diff
@@ -57,14 +58,84 @@ Global Variables
  * before transferring files */
 int axl_make_directories;
 
-/* AXL's flush file, SCR has one as well */
-char* axl_flush_file = NULL;
+/*
+ * Array for all the AXL_Create'd kvtree pointers.  It's indexed by the AXL id.
+ *
+ * Note: We only expand this array, we never shrink it.  This is fine since
+ * the user is only going to be calling AXL_Create() a handful of times.  It
+ * also simplifies the code if we never shrink it, and the extra memory usage
+ * is negligible, if any at all.
+ */
+kvtree **axl_kvtrees;
+static unsigned int axl_kvtrees_count = 0;
 
-/* Transfer handle unique IDs */
-static int axl_next_handle_UID = -1;
+static pthread_mutex_t id_lock = PTHREAD_MUTEX_INITIALIZER;
 
-/* tracks list of files written with transfer */
-kvtree* axl_file_lists = NULL;
+/*
+ * Allocate a new kvtree and return the AXL ID for it.  If state_file is
+ * specified, then populate the kvtree with it's data.
+ */
+static int axl_alloc_id(const char* state_file)
+{
+    kvtree *new;
+
+    new = kvtree_new();
+    if (state_file) {
+         /* initialize kvtree values from state_file if we have one */
+        if (access(state_file, F_OK) == 0 &&
+            kvtree_read_file(state_file, new) != KVTREE_SUCCESS) {
+            AXL_ERR("Couldn't read state file correctly\n");
+            return AXL_FAILURE;
```

Comment:
Let's return -1 here, since we've been using that value to represent a bad id.  There is no guarantee that AXL_FAILURE will always be defined as -1.  If someone decides to change its definition to be a positive value instead like 1, this will show up as a rare bug they might overlook.